master/internal/db/postgres_groups.go

```diff
@@ -0,0 +1,194 @@
+// group isn't found.
+func (db *PgDB) UpdateGroup(ctx context.Context, group model.Group) error {
+	res, err := Bun().NewUpdate().Model(&group).WherePK().Exec(ctx)
+
+	return matchSentinelError(mustHaveAffectedRows(res, err))
+}
+
+// AddUsersToGroup adds users to a group by creating GroupMembership rows.
+// Returns ErrNotFound if the group isn't found or ErrDuplicateRow if one
+// of the users is already in the group.
+func (db *PgDB) AddUsersToGroup(ctx context.Context, gid int, uids ...model.UserID) error {
+	if len(uids) < 1 {
+		return nil
+	}
+
+	groupMem := make([]model.GroupMembership, 0, len(uids))
+	for _, uid := range uids {
+		groupMem = append(groupMem, model.GroupMembership{
+			UserID:  uid,
+			GroupID: gid,
+		})
+	}
+
+	tx, err := Bun().BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	res, err := tx.NewInsert().Model(&groupMem).Exec(ctx)
+	if foundErr := mustHaveAffectedRows(res, err); foundErr != nil {
+		return matchSentinelError(foundErr)
+	}
+
+	err = tx.Commit()
+	return matchSentinelError(err)
+}
+
+// RemoveUsersFromGroup removes users from a group. Removes nothing and
+// returns ErrNotFound if the group or one of the users' membership rows
+// aren't found.
+func (db *PgDB) RemoveUsersFromGroup(ctx context.Context, gid int, uids ...model.UserID) error {
+	if len(uids) < 1 {
+		return nil
+	}
+
+	tx, err := Bun().BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	res, err := tx.NewDelete().Table("user_group_membership").
+		Where("group_id = ?", gid).
+		Where("user_id IN (?)", bun.In(uids)).
+		Exec(ctx)
+	if foundErr := mustHaveAffectedRows(res, err); foundErr != nil {
+		return matchSentinelError(foundErr)
+	}
+
+	err = tx.Commit()
+	return matchSentinelError(err)
+}
+
+// GetUsersInGroup searches for users that belong to a group and returns them.
+// Does not return ErrNotFound if none are found, as that is considered a
+// successful search.
+func (db *PgDB) GetUsersInGroup(ctx context.Context, gid int) ([]model.User, error) {
+	var users []model.User
+	err := Bun().NewSelect().Table("users").Model(&users).
+		Join("INNER JOIN user_group_membership AS ugm ON users.id=ugm.user_id").
+		Where("ugm.group_id = ?", gid).
+		Scan(ctx)
+
+	return users, err
+}
+
+func matchSentinelError(err error) error {
+	if errors.Is(err, sql.ErrNoRows) {
+		return ErrNotFound
+	}
+
+	switch pgErrCode(err) {
+	// Error code "foreign_key_violation"
+	case "23503":
+		return ErrNotFound
+	// Error code "unique_violation"
+	case "23505":
+		return ErrDuplicateRecord
+	}
+
+	return err
+}
+
+// mustHaveAffectedRows checks if bun has affected rows in a table or not.
+// Returns ErrNotFound if no rows were affected and returns the provided error otherwise.
+func mustHaveAffectedRows(result sql.Result, err error) error {
+	if err == nil {
+		rowsAffected, affectedErr := result.RowsAffected()
+		if affectedErr == nil && rowsAffected == 0 {
```

Comment:
Great question. AFAIK we should only get an error here when the driver doesn't support this operation. But on further consideration, since calling code might depend on this all-or-nothing behavior, I'm leaning toward returning `affectedErr`. I'll update the code to do so, and if anyone disagrees, please say so!